### PR TITLE
Expose autocast kwargs and simplify autocast wrapper

### PR DESCRIPTION
**Pull request by muellerzr**
This PR introduces `AutocastKwargs`, which help specify a few parameters to `torch.autocast` for the user to use. With this, we also change the internal logic in `Accelerator` to use `get_mixed_precision_context_manager` when grabbing the context manager, rather than duplicating the logic so these kwargs can be passed down. 

`enabled` is there so that users can do something such as:

```python
        kwargs = AutocastKwargs(enabled=False)
        AcceleratorState._reset_state()
        accelerator = Accelerator(mixed_precision="fp16")

        a_float32 = torch.rand((8, 8), device=accelerator.device)
        b_float32 = torch.rand((8, 8), device=accelerator.device)
        c_float32 = torch.rand((8, 8), device=accelerator.device)
        d_float32 = torch.rand((8, 8), device=accelerator.device)

        with accelerator.autocast():
            e_float16 = torch.mm(a_float32, b_float32)
            assert e_float16.dtype == torch.float16

            with accelerator.autocast(autocast_handler=kwargs):
                # Convert e_float16 to float32
                f_float32 = torch.mm(c_float32, e_float16.float())
                assert f_float32.dtype == torch.float32

            g_float16 = torch.mm(d_float32, f_float32)
            # We should be back in fp16
            assert g_float16.dtype == torch.float16
```

Solves https://github.com/huggingface/accelerate/issues/1737

### CODE REVIEW

**Review by sgugger — APPROVED** [review]
LGTM, thanks!

File: src/accelerate/utils/dataclasses.py [hunk 1]
```diff
@@ -51,6 +51,28 @@ def to_kwargs(self):
...
+    from accelerate import Accelerator
+    from accelerate.utils import AutocastKwargs
+
+    kwargs = AutocastKwargs(cache_enabled=True)
+    accelerator = Accelerator(kwargs_handlers=[kwargs])
+    ```
+    """
+
+    enabled: bool = False
+    cache_enabled: bool = False
```
Comment (yuxinyuan) [inline src/accelerate/utils/dataclasses.py:73]: Shouldn't the default value of ``cache_enabled`` be True, as that's the default value of pytorch

Comment (muellerzr) [reply src/accelerate/utils/dataclasses.py:73]: The default value is `None`, but yes. I will change this. Nice catch.

**Review by sgugger — APPROVED** [review]
Thanks!

**Pull request merged**